browser/ui/ai_chat/print_preview_extractor.cc

```diff
@@ -166,342 +167,383 @@ class PreviewPageTextExtractor {
   size_t total_page_count_ = 0;
   base::ReadOnlySharedMemoryRegion pdf_region_;
   base::OnceCallback<void(std::string)> callback_;
-  const uint32_t max_page_content_length_;
   mojo::Remote<printing::mojom::PdfToBitmapConverter> pdf_to_bitmap_converter_;
   base::WeakPtrFactory<PreviewPageTextExtractor> weak_ptr_factory_{this};
 };
 
-PrintPreviewExtractor::PrintPreviewExtractor(content::WebContents* web_contents,
-                                             Profile* profile,
-                                             bool is_pdf)
-    : web_contents_(web_contents),
-      profile_(profile),
-      active_chat_tab_helper_(AIChatTabHelper::FromWebContents(web_contents)),
-      is_pdf_(is_pdf) {
-  DCHECK(web_contents_);
-}
+class PrintPreviewExtractor : public printing::mojom::PrintPreviewUI {
+ public:
+  PrintPreviewExtractor(content::WebContents* web_contents,
+                        Profile* profile,
+                        bool is_pdf,
+                        AIChatTabHelper::ExtractPrintPreviewCallback callback)
+      : web_contents_(web_contents),
+        profile_(profile),
+        is_pdf_(is_pdf),
+        callback_(std::move(callback)) {
+    DCHECK(web_contents_);
+  }
 
-PrintPreviewExtractor::~PrintPreviewExtractor() {
-  ClearPreviewUIId();
-}
+  ~PrintPreviewExtractor() override { ClearPreviewUIId(); }
 
-mojo::PendingAssociatedRemote<PrintPreviewUI>
-PrintPreviewExtractor::BindPrintPreviewUI() {
-  return print_preview_ui_receiver_.BindNewEndpointAndPassRemote();
-}
+  PrintPreviewExtractor(const PrintPreviewExtractor&) = delete;
+  PrintPreviewExtractor& operator=(const PrintPreviewExtractor&) = delete;
 
-void PrintPreviewExtractor::DisconnectPrintPrieviewUI() {
-  print_preview_ui_receiver_.reset();
-}
+  void CreatePrintPreview() {
+    if (profile_->GetPrefs()->GetBoolean(::prefs::kPrintPreviewDisabled)) {
+      DeleteSelfAndSendResult("");
+      return;
+    }
+    content::RenderFrameHost* rfh = GetRenderFrameHostToUse(web_contents_);
+    if (rfh) {
+      if (!print_render_frame_.is_bound()) {
+        rfh->GetRemoteAssociatedInterfaces()->GetInterface(
+            &print_render_frame_);
```

Comment:
Why are we changing this class to be self deleted? `PrintPreviewExtractor` was owned by AIChatUI to ensure it won't outlive the chat panel and also ensure there is only one instance exists at a time per renderer because `mojo::AssociatedRemote<printing::mojom::PrintRenderFrame> print_render_frame_` can only be bound to one receiver at the same time. Self deleting will break that guarantee.

Reply:
That isn't the only problem here, it could also outlive the webcontents

Reply:
the webcontents issue could be fixed by passing the frame tree node id and using it to get the webcontents each time, but there may be other issues as @darkdh mentioned

Reply:
It also seems very easy to accidentally leak this class if there's some code path that is missing the call to delete itself.

Reply:
>  was owned by AIChatUI to ensure it won't outlive the chat panel 

As explained, Conversations are now decoupled from `AIChatUI`. You can close a conversation in a panel, mid-generation, and pick up on it in the full-page, or sometime later. Any `AIChatUI` can bind to any conversation from any `WebContents` (or standalone conversations but that's not relevant here). In fact, `AIChatUI` doesn't know which conversation is open in the UI, the renderer binds to the `ConversationHandler` directly. Not to mention the previous pattern was very confusing IMO since it used an observer, but we want to guarantee that only 1 observer is attached (which it wasn't). It is a 1:1 relationship purely so that we can run code in `chrome/browser/ui` so some kind of delegate or provider pattern makes more sense.

> ensure there is only one instance exists at a time per renderer because mojo::AssociatedRemote<printing::mojom::PrintRenderFrame> print_render_frame_ can only be bound to one receiver at the same time

What guarantee was there previously in `PrintPreviewExtractor` that isn't there now? In this PR, `ConversationDriver::GetContent` is guarded by `is_page_text_fetch_in_progress_` and that's the only place which calls `AIChatTabHelper::GetPageContent` which is the only place which runs the print preview extractor. All page content fetching is guaranteed to only happen once.

> it could also outlive the webcontents

Yes that's an issue. This pattern is also used in `FetchPageContent`, and in Chromium's `FetchRssLinks`, which is admittedly simpler and uses a mojom disconnect handler to `SendResultAndDelete` when the `WebContents` is destroyed. In this case, we could abandon when the `WebContents` is destroyed because the callback to the `AIChatTabHelper` wouldn't be valid anymore anyway. I suppose I'll change this to be yet another delegate. It has to be a delegate and not directly owned because `AIChatTabHelper` is in `components/ai_chat/content/browser`, but `PrintPreviewExtractor` requires access to `chrome/browser` and `chrome/browser/ui`. It seems less convenient to move `AIChatTabHelper` to `brave/browser/ui`.

Reply:
> What guarantee was there previously in PrintPreviewExtractor that isn't there now? 
```  
print_preview_extractor_ = std::make_unique<PrintPreviewExtractor>(
      active_chat_tab_helper_->web_contents(), profile_, is_pdf);
```
To make sure there is only one instance per web_contents disregarding previous one is still running or not.

Reply:
This is addressed now. AIChatTabHelper is provided with an `AIChatTabHelper::PrintPreviewExtractionDelegate` instance, `PrintPreviewExtractor`, which itself owns an instance of `PrintPreviewExtractorInternal`, which it creates and resets on each request to `Extract()`, to match the current implementation of construction for each extraction request.

Reply:
The new delegate approach is neater than the self deleted one and even original subclassing `printing::mojom::PrintPreviewUI` in Leo panel UI.